In [ ]:
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense


X = np.array([
    [0,0],
    [0,1],
    [1,0],
    [1,1]
])

y = np.array([
    [0],
    [1],
    [1],
    [0]
])
model=Sequential([
    Dense(4,activation='tanh',input_shape=(2,)),
    Dense(1,activation='sigmoid')
])

model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
model.fit(X, y,

          epochs=1000, 
          verbose=1,
          
)

Epoch 1/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 597ms/step - accuracy: 0.5000 - loss: 0.7643
Epoch 2/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - accuracy: 0.5000 - loss: 0.7637
Epoch 3/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - accuracy: 0.5000 - loss: 0.7632
Epoch 4/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.5000 - loss: 0.7626
Epoch 5/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step - accuracy: 0.5000 - loss: 0.7621
Epoch 6/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.5000 - loss: 0.7615
Epoch 7/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - accuracy: 0.5000 - loss: 0.7610
Epoch 8/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.5000 - loss: 0.7604
Epoch 9/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.5000 - loss: 0.7599
Epoch 10/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step - accuracy: 0.5000 - loss: 0.7594
Epoch 11/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step - accuracy: 0.5000 - loss: 0.7588
Epoch 12/1000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - accuracy

In [12]:
loss, accuracy = model.evaluate(X, y)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 141ms/step - accuracy: 0.7500 - loss: 0.5497


In [14]:
pred=model.predict(X)
print(f"  Final Loss     : {loss:.6f}")
print(f"  Final Accuracy : {accuracy * 100:.1f}%")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step
  Final Loss     : 0.549714
  Final Accuracy : 75.0%


# XOR Neural Network

## Forward Propagation
During forward propagation, input data (XOR pairs [0,0], [0,1], [1,0], [1,1]) flows through the network's layers. The inputs first reach the hidden layer with 4 neurons using tanh activation, which transforms the data through weighted connections. These transformed values then pass to the output layer with a single neuron using sigmoid activation, producing predictions between 0 and 1 for each input pattern.

## Error Computation
The network computes error by comparing its predictions against the true XOR outputs ([0,1,1,0]) using binary cross-entropy loss. This loss function quantifies the difference between predicted and actual values, with higher values indicating worse performance. As shown in the training output, the loss starts at 0.7643 and gradually decreases to 0.5500 by epoch 1000.

## Backpropagation
In backpropagation, the network calculates how each weight contributed to the error by applying the chain rule backwards through the network layers. It determines the gradient of the loss with respect to every weight, identifying which weights need adjustment and in which direction to reduce the overall error.

## Weight Updates
The Adam optimizer updates the network weights after each epoch using the calculated gradients. Weights are adjusted in small steps to minimize the error, which explains the steady loss decrease from 0.7643 to 0.5500 over 1000 epochs. Despite these updates, the network only achieved 75% accuracy, successfully learning 3 of the 4 XOR patterns.